[Week5][Davidyang2149] Add Search Checkbox for Restaurants info - 구현 진행 중

DavidYang2149 opened the pull request:
# 5강 비동기 - 5주 차 과제 1
## 기능 구현
- [ ] TDD로 기능 구현
- [ ] 레스토랑 지역 목록 보기 기능
- [ ] 레스토랑 카테고리 목록 보기 기능
- [ ] 레스토랑 목록 보기 기능
- [ ] 커버리지 100% 달성

## 제한 조건
- [ ] 모든 기능은 TDD로 구현합니다. 컴포넌트와 reducer에 대한 테스트가 존재해야 합니다. 먼저 테스트를 추가하고 commit 한 뒤 실제 기능을 추가하고 commit합니다. commit 메시지 로그는 아래와 같이 남아야 합니다.
```sh
Restaurant에 대한 테스트를 추가한다
Restaurant을 구현한다
Restaurant를 리팩토링한다
```
- [ ] if는 사용할 수 있어도 else 사용하실 수 없습니다. GuardClauses 방법을 사용해주세요.
- [ ] switch는 사용하실 수 없습니다.
- [ ] let은 사용하실 수 없습니다. const만을 사용하여 문제를 해결해주세요.
- [ ] 함수 이름과 변수 이름에 줄임말은 사용하실 수 없습니다. 길더라도 명확하게 써주세요.
- [ ] indent(인덴트, 들여쓰기) depth를 1로 유지해주세요. 예를 들어, for문 안에 if문이 있으면 indent depth는 2입니다. depth를 줄이는 좋은 방법은 함수(또는 메소드)를 분리하면 됩니다.

## 요구사항
- [ ] ESLint를 돌린 결과 아무런 문제가 없어야 합니다.
- [ ] 모든 인수 테스트를 통과시켜야 합니다.
- [ ] 한 파일에는 하나의 컴포넌트만 있어야 합니다.
- [ ] 테스트 커버리지 100%를 달성해야 합니다.

## 작업내용 및 질의
- [ ] 구현 작업 진행 중

DavidYang2149 committed:
[Red] Add App.test.jsx for TDD

5강 비동기 - 5주 차 과제 1
- App.test.jsx 추가.
- App 테스트 코드 작성(Red).

DavidYang2149 committed:
[Green] Update App.jsx for Implement

5강 비동기 - 5주 차 과제 1
- App.jsx 업데이트.
- 테스트 코드 작성 후 UI 구현.

hannut91 commented:
리팩터링을 할 차례군요 ㅎㅎ 리팩터링은 꼭 빼먹으시면 안됩니다~!

DavidYang2149 committed:
[Refactor] Update App.jsx and App.test.jsx

5강 비동기 - 5주 차 과제 1
- App.jsx 및 App.test.jsx 리팩토링.
- 레스토랑 검색결과 Unorder List 생성.

DavidYang2149 committed:
[Red] Add Restaurants.test.jsx for TDD

5강 비동기 - 5주 차 과제 1
- Restaurants.test.jsx 추가.
- Restaurants 테스트 코드 작성(Red).

DavidYang2149 committed:
[Green] Add Restaurants.jsx for Implement

5강 비동기 - 5주 차 과제 1
- Restaurants.jsx 추가.
- 테스트 코드 작성 후 UI 구현(Green).

DavidYang2149 committed:
[Refactor] Update Restaurants.jsx

5강 비동기 - 5주 차 과제 1
- Restaurants.jsx 리팩토링.
- 레스토랑 배열 생성.

DavidYang2149 committed:
[Refactor] Update App.jsx

5강 비동기 - 5주 차 과제 1
- App.jsx 리팩토링.
- Restaurants 컴포넌트 수신.

hannut91 commented:
바쁠수록 공부할 것이 많을수록 얼마나 공부할 것인가를 잘 정해야합니다. 과제를 전부 안하셔도 괜찮아요 ㅎㅎ 하지만 시간을 정해놓고 할애하셨으면 좋겠어요.

DavidYang2149 committed:
[Red] Add Regions.test.jsx for TDD

5강 비동기 - 5주 차 과제 1
- Regions.test.jsx 추가.
- Regions 테스트 코드 작성(Red).

DavidYang2149 committed:
[Green] Update Regions.jsx for Implement

5강 비동기 - 5주 차 과제 1
- Regions.jsx 업데이트.
- 테스트 코드 작성 후 UI 구현.

DavidYang2149 committed:
[Refactor] Update App.jsx

5강 비동기 - 5주 차 과제 1
- App.jsx 리팩토링.
- Regions 컴포넌트 수신.

DavidYang2149 committed:
[Red] Add Categories.test.jsx for TDD

5강 비동기 - 5주 차 과제 1
- Categories.test.jsx 추가.
- Categories 테스트 코드 작성(Red).

DavidYang2149 committed:
[Green] Add Categories.jsx for Implement

5강 비동기 - 5주 차 과제 1
- Categories.jsx 업데이트.
- 테스트 코드 작성 후 UI 구현.

DavidYang2149 committed:
[Refactor] Update App.jsx

5강 비동기 - 5주 차 과제 1
- App.jsx 리팩토링.
- Categories 컴포넌트 수신.

DavidYang2149 committed:
[Add] __mocks__/ & __fixtures__/ for TDD

5강 비동기 - 5주 차 과제 1
- 테스트 케이스 작성을 위한 목업 작성.

DavidYang2149 committed:
[Refactor] Update Categories.jsx & Regions.jsx & Restaurants.jsx

5강 비동기 - 5주 차 과제 1
- Categories.jsx & Regions.jsx & Restaurants 리팩토링.
- 각 컴포넌트 리스트의 key 생성.

DavidYang2149 committed:
[Red] Update Regions.test.jsx for TDD

5강 비동기 - 5주 차 과제 1
- Regions.test.jsx 수정.
- Regions 테스트 코드 작성(Red).

DavidYang2149 committed:
[Red] Update Regions.test.jsx for TDD

5강 비동기 - 5주 차 과제 1
- Regions.test.jsx 수정.
- Regions 테스트 코드 작성(Red).

DavidYang2149 committed:
[Green] Update Regions.jsx

5강 비동기 - 5주 차 과제 1
- Regions.jsx 수정.
- 테스트 코드 작성 후 UI 구현(버튼).

DavidYang2149 committed:
[Refactor] Update App.jsx

5강 비동기 - 5주 차 과제 1
- App.jsx 리팩토링.
- Regions handleClick 추가.

DavidYang2149 committed:
[Add] Folder add(modules) initialize for Redux

5강 비동기 - 5주 차 과제 1
- modeuls 폴더 생성.
- 리덕스(actoins.js & reducer.js & store.js) 구성 추가.

DavidYang2149 committed:
[Update] Update index.jsx for Provider

5강 비동기 - 5주 차 과제 1
- index.jsx 업데이트.
- Provider 추가.

hannut91 commented:
TDD연습을 잘해주시고 계시네요 ㅎㅎ

DavidYang2149 committed:
[Update] Folder update(modules) Redux

5강 비동기 - 5주 차 과제 1
- modeuls 폴더 업데이트.
- 리덕스(actoins.js & reducer.js & store.js) 구성 추가.

DavidYang2149 committed:
[Red] Add RegionsContainer.test.jsx for TDD

5강 비동기 - 5주 차 과제 1
- RegionsContainer.test.jsx 생성.
- RegionsCpmaomer 테스트 코드 작성(Red).

DavidYang2149 committed:
[Green] Add RegionsContainer.jsx

5강 비동기 - 5주 차 과제 1
- RegionsContainer.jsx 생성.
- 컨테이너 생성 및 로직 구현.

DavidYang2149 committed:
[Refactor] Update RegionsContainer.jsx

5강 비동기 - 5주 차 과제 1
- RegionsContainer.jsx 수정.
- 기능 및 변수 추가.

DavidYang2149 committed:
[Refactor] Update Regions.jsx

5강 비동기 - 5주 차 과제 1
- Regions.jsx 수정.
- 기능 및 변수 추가.

DavidYang2149 committed:
[Refactor] Update App.jsx

5강 비동기 - 5주 차 과제 1
- App.jsx 수정.
- api 데이터 수신으로 변경.

DavidYang2149 committed:
[Add] File service/ add for api

5강 비동기 - 5주 차 과제 1
- services 파일 생성.
- api 생성.

DavidYang2149 committed:
[Update] update File modules

5강 비동기 - 5주 차 과제 1
- Regions 컴포넌트용 redux 구조 수정.
- 기능 추가.

DavidYang2149 committed:
[Red] Update RegionsContainer.test.jsx

5강 비동기 - 5주 차 과제 1
- RegionsContainer.test.jsx 수정.
- (문제) 구현 순서가 뒤바뀜.
- Green 먼저 구현 후 Red 실행.

DavidYang2149 committed:
[Red] Update Regions.test.jsx

5강 비동기 - 5주 차 과제 1
- Regions.test.jsx 수정.
- (문제) 구현 순서가 뒤바뀜.
- Green 먼저 구현 후 Red 실행.

DavidYang2149 committed:
[Red] Update App.test.jsx

5강 비동기 - 5주 차 과제 1
- App.test.jsx 수정.
- (문제) 구현 순서가 뒤바뀜.
- Green 먼저 구현 후 Red 실행.

DavidYang2149 committed:
[Red] Add CategoriesContainer.test.jsx for TDD

5강 비동기 - 5주 차 과제 1
- CategoriesContainer.test.jsx 생성.
- RegionsContainer.test.jsx를 참조하여 테스트 코드 작성(Red).

DavidYang2149 committed:
[Green] Add CategoriesContainer.jsx

5강 비동기 - 5주 차 과제 1
- CategoriesContainer.jsx 생성.
- 컨테이너 생성 및 로직 구현.
- RegionsContainer.jsx 참조.

DavidYang2149 committed:
[Red] Update Categories.test.jsx

5강 비동기 - 5주 차 과제 1
- Categories.test.jsx 수정.
- Categories 테스트 코드추가 작성(Red).
- CategoriesContainer.jsx 연결.

DavidYang2149 committed:
[Green] Update Categories.jsx

5강 비동기 - 5주 차 과제 1
- Categories.jsx 수정.
- CategoriesContainer.jsx 적용에 따른 수정.

DavidYang2149 committed:
[Red] Update App.test.jsx

5강 비동기 - 5주 차 과제 1
- App.test.jsx 수정.
- App 테스트 코드 추가 작성(Red).
- CategoriesContainer.jsx 추가.

DavidYang2149 committed:
[Green] Update App.jsx

5강 비동기 - 5주 차 과제 1
- App.jsx 수정.
- CategoriesContainer.jsx 추가.

DavidYang2149 committed:
[Update] Update Folder modules Redux for Categories

5강 비동기 - 5주 차 과제 1
- modules 수정.
- 리덕스(Categories) 구성 추가.

DavidYang2149 committed:
[Update] Update File service

5강 비동기 - 5주 차 과제 1
- services 파일 수정.
- api 추가(fetchRestaurants).

DavidYang2149 committed:
[Red] Add RestaurantsContainer.test.jsx for TDD

5강 비동기 - 5주 차 과제 1
- RestaurantsContainer.test.jsx 생성.
- CategoriesContainer.test.jsx를 참조하여 테스트 코드 작성(Red).

DavidYang2149 committed:
[Green] Add RestaurantsContainer.jsx

5강 비동기 - 5주 차 과제 1
- RestaurantsContainer.jsx 생성.
- 컨테이너 생성 및 로직 구현.
- CategoriesContainer.jsx 참조.

DavidYang2149 committed:
[Red] Update Restaurants.test.jsx for TDD

5강 비동기 - 5주 차 과제 1
- Restaurants.test.jsx 수정.
- Restaurants 추가 테스트 코드 작성(Red).

DavidYang2149 committed:
[Green] Update Restaurants.jsx

5강 비동기 - 5주 차 과제 1
- Restaurants.jsx 수정.
- 테스트 코드 작성 후 로직 변경(Green).

DavidYang2149 committed:
[Update] Folder update(modules) Redux

5강 비동기 - 5주 차 과제 1
- modeuls 폴더 수정.
- 리덕스(restaurants) 구성 추가.

DavidYang2149 committed:
[Red] Update App.test.jsx

5강 비동기 - 5주 차 과제 1
- App.test.jsx 수정.
- RestaurantsContainer.jsx 수용.

DavidYang2149 committed:
[Green] Update App.jsx

5강 비동기 - 5주 차 과제 1
- App.jsx 수정.
- Restaurants.jsx 수용.

DavidYang2149 committed:
[Refactor] Update RegionsContainer.jsx

5강 비동기 - 5주 차 과제 1
- RegionsContainer.jsx 수정.
- getRestaurants 수용 및 일부 명칭 수정.

DavidYang2149 committed:
[Refactor] Update Regions.jsx

5강 비동기 - 5주 차 과제 1
- Regions.jsx 수정.

DavidYang2149 committed:
[Refactor] Update Regions.test.jsx

5강 비동기 - 5주 차 과제 1
- Regions.test.jsx 수정.
- 명칭 수정.

DavidYang2149 committed:
[Red] Update CategoriesContainer.test.jsx

5강 비동기 - 5주 차 과제 1
- CategoriesContainer.test.jsx 수정.
- Restaruants 추가에 따른 테스트 코드 작성(Red).

DavidYang2149 committed:
[Green] Update CategoriesContainer.jsx

5강 비동기 - 5주 차 과제 1
- CategoriesContainer.jsx 수정.
- Restaurants 수용 구현.

DavidYang2149 committed:
[Red] Update Categories.test.jsx

5강 비동기 - 5주 차 과제 1
- Categories.test.jsx 수정.
- Categories 테스트 코드추가 작성(Red).

DavidYang2149 committed:
[Green] Update Categories.jsx

5강 비동기 - 5주 차 과제 1
- Categories.jsx 수정.
- 로직 수정.

DavidYang2149 committed:
[Add] api.test.js Add

5강 비동기 - 5주 차 과제 1
- api.test.js 추가.

hannut91 commented on src/Categories.jsx:
```diff
@@ -0,0 +1,17 @@
+        {categories.map(({ id, name }) => (
+          <li key={id}>
+            <button type="button" onClick={() => onClick(id)}>
+              {selected === id ? `${name}(V)` : name}
```
name과 선택됨을 표시하는 것을 분리하면 좋을 것 같습니다.

```js
{name}
{selected && '(V')}
```

이렇게 하다보면 선택됨을 표시하는 것을 컴포넌트로 분리해도 되겠죠

hannut91 commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,60 @@
+  const handleClick = jest.fn();
+
+  it('Page render without selected', () => {
+    const selected = '';
```
selected라는 이름만 보면 true 혹은 false만 반환할 것 같아요. 하지만 실제로는 지금 빈 문자열로 초기화 해주고 있고, 또 밑에서는 숫자로 사용하고 있어요. 적절한 이름을 사용하던지 아니면 자료형에대해서 더 고민해보시면 좋을 것 같아요

hannut91 commented on src/Regions.test.jsx:
```diff
@@ -0,0 +1,62 @@
+    expect(getByText('서울(V)')).not.toBeNull();
+  });
+
+  it('Click region', () => {
```
실제로 어떤 일이 일어나는지, 어떤 변화가 일어나는지에 대해서 작성하면 좋을 것 같아요

DavidYang2149 committed:
[Add] actions.test.js

5강 비동기 - 5주 차 과제 1
- actions.test.js 추가.

DavidYang2149 commented:
actions.test.js 구현을 실패하였습니다.
문제를 해결해보고자 했으나 원인을 찾지 못하였습니다.

hannut91 commented:
비동기 액션 같은 경우에는 따로 테스트하는게 좋을 것 같아요. 

예를들어서 레스토랑 목록을 가져오는 것을 볼게요.

```js
export const loadRestaurants = () => async (dispatch, getState) => {
  const restaurants = await restaurantService.getRestaurants();
  
  dispatch(setRestaurants(restaurants));
}
```

이 비동기 액션에서 하는 일은

1. 서버에서 레스토랑 목록을 가져온다.

2. 가져온 레스토랑 목록을 setRestaurants액션을 통해 dispatch한다.

restaurantService에서 레스토랑 목록을 요청한 후 레스토랑을 업데이트 하는 setRestauarnts라는 액션을 다시 dispatch합니다. 하지만 레스토랑 목록을 어떻게 가져오는지는 주요한 관심사가 아닙니다. 그래서 getRestaurants는 mocking하여 처리하고 올바르게 가져왔다고 가정하고 액션이 올바르게 처리되는지 테스트합니다.

```js
import thunk from 'redux-thunk';

import configureStore from 'redux-mock-store';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const restaurants = [
  { id: 1, name: '마녀주방', category: '한식', address: '강남' },
];

getRestaurants.mockResvoledValue(restaurants);

test('loadRestaurants', async () => {
  store = mockStore({});
  
  await loadRestaurants();
  
  const actions = store.getActions();
  
  expect(actions[0]).toEqual(setRestaurants(restaurants));
});
```

See also
  - https://github.com/reduxjs/redux-mock-store

DavidYang2149 commented:
개념이 이해가 되지 않아 천천히 반복하며 학습해보도록 하겠습니다!
